### ISSUE TITLE
Reproducing PCA results

### ISSUE THREAD

**Issue by cstjean** [madeleineudell/LowRankModels.jl#53]
I can't reproduce PCA outcomes from other libraries with glrm. Am I doing something wrong? 
```julia
using LowRankModels: pca, fit!, ProxGradParams
using RDatasets: dataset

iris = dataset("datasets", "iris")
norm_rows(mat) = mat ./ sqrt(sum(mat .^ 2, 2))
X_iris = convert(Matrix, iris[[:SepalLength, :SepalWidth, :PetalLength, :PetalWidth]])
srand(33)
glrm_pca = pca(X_iris, 2)
fit!(glrm_pca, ProxGradParams(max_iter=1000, convergence_tol=1.e-20))
norm_rows(glrm_pca.Y)

> 2x4 Array{Float64,2}:
>   0.802641   0.560961   0.201397   0.0229959
>  -0.622311  -0.187197  -0.711173  -0.268178 
```
Both scikit-learn-python and MultivariateStats.jl return
```
2x4 Array{Float64,2}:
  0.361387  -0.0845225  0.856671  0.358289
 -0.656589  -0.730161   0.173373  0.075481
```

**Comment by cstjean** [comment]
This is on `dataframe-ux`, for what it's worth.

**Comment by madeleineudell** [comment]
The PCA objective is invariant under a very large class of transformations,
since X'*Y = (QX)' * (Q^{-1}Y) for any invertible matrix Q.

scikit-learn-python and MultivariateStats are both probably computing PCA
via the singular value decomposition: to approximate A, they find A = U S
V^T and then return S^{1/2} V^T as the factor Y. LowRankModels doesn't take
advantage of the analytic structure of PCA, but just performs optimization
on the objective exactly like it would for any other objective. So it might
find a solution corresponding to any transformation Q, since they all give
exactly the same objective value.

To check if the packages are giving equally good answers, it's best to look
at X'*Y or the objective value.

Madeleine

On Tue, Apr 26, 2016 at 8:56 AM, Cédric St-Jean <notifications@github.com>
wrote:

> This is on dataframe-ux, for what it's worth.
>
> —
> You are receiving this because you are subscribed to this thread.
> Reply to this email directly or view it on GitHub
> <https://github.com/madeleineudell/LowRankModels.jl/issues/53#issuecomment-214792620>
>



-- 
Madeleine Udell
Postdoctoral Fellow at the Center for the Mathematics of Information
California Institute of Technology
*https://courses2.cit.cornell.edu/mru8
<https://courses2.cit.cornell.edu/mru8>*
(415) 729-4115

**Comment by cstjean** [comment]
I just checked, and the `X'*Y` from ScikitLearn and from GLRM are 5 times closer to each other in root-mean-squared-distance than they are to `X_iris`, so you're definitely right. I should have though about that, thank you!

**Comment by madeleineudell** [comment]
It's a common misunderstanding, unfortunately. People often interpret the
factors they get from PCA without realizing that the interpretations
they're making are *not* invariant under transformations Q and therefore
bear almost no relation to the data...

On Tue, Apr 26, 2016 at 11:26 AM, Cédric St-Jean <notifications@github.com>
wrote:

> Closed #53 <https://github.com/madeleineudell/LowRankModels.jl/issues/53>.
>
> —
> You are receiving this because you commented.
> Reply to this email directly or view it on GitHub
> <https://github.com/madeleineudell/LowRankModels.jl/issues/53#event-641897776>
>



-- 
Madeleine Udell
Postdoctoral Fellow at the Center for the Mathematics of Information
California Institute of Technology
*https://courses2.cit.cornell.edu/mru8
<https://courses2.cit.cornell.edu/mru8>*
(415) 729-4115
